scheds/rust/scx_layered/src/main.rs

```diff
@@ -1246,6 +1268,211 @@ impl Layer {
         Ok(nr_new_cpus)
     }
 }
+#[derive(Debug, Clone)]
+struct NodeInfo {
+    node_mask: nix::sched::CpuSet,
+    _node_id: usize,
+}
+#[derive(Debug)]
+struct GpuTaskHandler {
+    gpu_devs_to_node_info: HashMap<u32, NodeInfo>,
+    gpu_pids_to_devs: HashMap<Pid, u32>,
+    last_process_time: Option<Instant>,
+    sys: System,
+    pid_map: HashMap<Pid, Vec<Pid>>,
+    poll_interval: Duration,
+    enable: bool,
+}
+
+impl GpuTaskHandler {
+    pub fn new(poll_interval: u64, enable: bool) -> GpuTaskHandler {
+        GpuTaskHandler {
+            gpu_devs_to_node_info: HashMap::new(),
+            gpu_pids_to_devs: HashMap::new(),
+            last_process_time: None,
+            sys: System::default(),
+            pid_map: HashMap::new(),
+            poll_interval: Duration::from_secs(poll_interval),
+            enable,
+        }
+    }
+    fn find_one_cpu(&self, affinity: Vec<u64>) -> Result<u32, anyhow::Error> {
+        let mut cpu: u32 = 0;
+        for &mask in affinity.iter() {
+            let mut inner_offset: u64 = 1;
+            for _ in 0..64 {
+                if (mask & inner_offset) != 0 {
+                    return Ok(cpu);
+                }
+                inner_offset = inner_offset << 1;
+                cpu += 1;
+            }
+        }
+        anyhow::bail!("unable to get CPU from NVML bitmask");
+    }
+    fn node_to_cpuset(&self, node: &scx_utils::Node) -> Result<CpuSet, anyhow::Error> {
+        let mut cpuset = CpuSet::new();
+        for (cpu_id, _cpu) in &node.all_cpus {
+            cpuset.set(cpu_id.clone())?;
+        }
+        Ok(cpuset)
+    }
+    fn init_dev_node_map(&mut self, topo: Arc<Topology>) -> Result<(), anyhow::Error> {
+        let nvml = nvml()?;
+        let device_count = nvml.device_count()?;
+
+        for idx in 0..device_count {
+            let dev = nvml.device_by_index(idx)?;
+            let cpu = dev.cpu_affinity(16)?;
+            let ideal_cpu = self.find_one_cpu(cpu)?;
+            for (node_id, node) in topo.nodes.iter() {
+                if node.all_cpus.contains_key(&(ideal_cpu as usize)) {
+                    self.gpu_devs_to_node_info.insert(
+                        idx,
+                        NodeInfo {
+                            node_mask: self.node_to_cpuset(node)?,
+                            _node_id: *node_id,
+                        },
```

Comment:
If we add a helper to return `CpuSet` to `Cpumask`, the above would just need to store gpu -> node_id, right?